Add Select operation

halfcrazy opened the pull request:
Fix https://github.com/ovn-kubernetes/libovsdb/issues/128

## `Select` Usage Changes

The `ConditionalAPI.Select()` method now generates the OVSDB `select` operation but does not execute the query itself. Callers must pass the generated `Operation` to `client.Transact()` for execution and then use `client.ParseSelectResult()` to parse the results.

**Core Workflow:** `WhereXxx(...).Select() -> Transact(...) -> ParseSelectResult(...)`

**Main Use Cases:**

1.  **Select All Rows:**
    Use `client.Where(&Model{})` (with a zero-value model instance) to set the table context, then call `.Select()`.
    ```go
    // var ovs client.Client
    // var ctx context.Context
    // type Bridge struct { ... } // Your model definition
    var allBridges []Bridge
    // 1. Generate Op: Where(&Bridge{}) specifies the table
    selectOp, _ := ovs.Where(&Bridge{}).Select()
    // 2. Transact
    reply, _ := ovs.Transact(ctx, selectOp...)
    // 3. Parse Result
    _ = ovs.ParseSelectResult(reply[0], &allBridges)
    ```

2.  **Select by AND Conditions:**
    Use `client.WhereAll(&Model{}, condition1, condition2, ...)` to set the table context and filter conditions (combined with AND), then call `.Select()`.
    ```go
    // var ovs client.Client
    // var ctx context.Context
    // type Bridge struct { Name string; ... }
    var specificBridges []Bridge
    cond := model.Condition{Field: &(&Bridge{}).Name, Function: ovsdb.ConditionEqual, Value: "br0"}
    // 1. Generate Op: Use WhereAll for one or more AND conditions
    selectOp, _ := ovs.WhereAll(&Bridge{}, cond).Select()
    // 2. Transact
    reply, _ := ovs.Transact(ctx, selectOp...)
    // 3. Parse Result
    _ = ovs.ParseSelectResult(reply[0], &specificBridges)
    ```

**Limitations:**

*   `WhereAny(...).Select()` is **not supported** (OVSDB `select` does not support OR conditions). Calling `.Select()` after `WhereAny` will return an error.
*   `WhereCache(...).Select()` is **not supported** (cache predicate functions cannot be translated to OVSDB conditions). Calling `.Select()` after `WhereCache` will return an error.

halfcrazy commented:
ping @dave-tucker

jcaamano commented:
Hello @halfcrazy 

Sorry with the delay. I had this PR on my radar and I think it's important so thanks for the effort. I have just been extremely busy but I will try to keep up actively reviewing it.

Before getting into the weeds and looking into the code, I would like to discuss some of the design decisions reflected or that would need to be reflected in the PR OP message :

1. `WhereAny`

From the operation perspective on the protocol, no operation supports `OR` conditions. This is achieved in the implementation splitting each element of the `OR` condition on its own operation. So if we had `whereAny(A OR B)`, then that would translate into two ops, one for A and another one for B. Results could then be aggregated.

I don't see any reason this wouldn't work for `Select` operation, or more specifically, I don't see the protocol as a reason not to implement it unless I am missing something.

2. `WhereCache`

I understand how `WhereCache` might imply the results come form the cache, but it could be used in a mode where we are monitoring some but not all the columns of a table.  So the predicates could on the cache, turn matches into UUID based conditions with which `Select` could be used to fetch non-monitored columns.

Again this is how all `ConditionalAPI`  operations run implicitly, and I don't see the protocol as a reason why it couldn't be the same for `Select`.

3. `Where`

You don't mention in the PR OP message as this being limited, but in practicality it would be limited for the same reasons `WhereAny` and `WhereCache` are limited so I would expect this to be supported/not supported and mentioned for the same reasons.

4.. `ParseSelectResult`

I think trying to keep operations and results as abstract to the user as we can is best. Having the user assume that a select translates to a single operation of known index is something that is already not true for other operations and would not allow us to do what I explained in (1). 

I would rather have something akin to the existing `func CheckOperationResults(result []OperationResult, ops []Operation) ([]OperationError, error)` but maybe more like `func GetSelectResults(ops []Operation, [][]model.Model) ([]OperationError, error)`, a user would pass a two dimensional slice instance, the first dimension corresponds to each select operation bundled in the transaction, and the second one is the result of each select operation. The logic should be able to correlate it from the original ops.


I would also update the README reflecting how to use this new op.

halfcrazy commented:
> Hello @halfcrazy
> 
> Sorry with the delay. I had this PR on my radar and I think it's important so thanks for the effort. I have just been extremely busy but I will try to keep up actively reviewing it.
> 
> Before getting into the weeds and looking into the code, I would like to discuss some of the design decisions reflected or that would need to be reflected in the PR OP message :
> 
> 1. `WhereAny`
> 
> From the operation perspective on the protocol, no operation supports `OR` conditions. This is achieved in the implementation splitting each element of the `OR` condition on its own operation. So if we had `whereAny(A OR B)`, then that would translate into two ops, one for A and another one for B. Results could then be aggregated.
> 
> I don't see any reason this wouldn't work for `Select` operation, or more specifically, I don't see the protocol as a reason not to implement it unless I am missing something.
> 
> 2. `WhereCache`
> 
> I understand how `WhereCache` might imply the results come form the cache, but it could be used in a mode where we are monitoring some but not all the columns of a table. So the predicates could run on the cache, turn matches into UUID based conditions with which `Select` could be used to fetch non-monitored columns.
> 
> Again this is how all `ConditionalAPI` operations run implicitly, and I don't see the protocol as a reason why it couldn't be the same for `Select`.
> 
> 3. `Where`
> 
> You don't mention in the PR OP message as this being limited, but in practicality it would be limited for the same reasons `WhereAny` and `WhereCache` are limited so I would expect this to be supported/not supported and mentioned for the same reasons.
> 
> 4.. `ParseSelectResult`
> 
> I think trying to keep operations and results as abstract to the user as we can is best. Having the user assume that a select translates to a single operation of known index is something that is already not true for other operations and would not allow us to do what I explained in (1).
> 
> I would rather have something akin to the existing `func CheckOperationResults(result []OperationResult, ops []Operation) ([]OperationError, error)` but maybe more like `func GetSelectResults(ops []Operation, [][]model.Model) ([]OperationError, error)`, a user would pass a two dimensional model slice instance, the first dimension corresponds to each select operation bundled in the transaction, and the second one is the result of each select operation. The logic should be able to correlate it from the original ops. **edit:** or maybe `func GetSelectResults(ops []Operation, index int, []<ModelType>) ([]OperationError, error)`
> 
> I would also update the README reflecting how to use this new op.

Thank you for reviewing. You're right. Because the transact operation is atomic, in theory, we can convert `OR` operations that are not supported by the server into multiple select statements on the client side. For operations like `WhereCache`, they can be converted into multiple select matches based on UUIDs. (Here, I have a concern about performance: the select statements generated by the chained `WhereCache` may likely form a huge transaction to submit.) The same logic applies to `Where` operations similar to `WhereCache`.

halfcrazy commented:
About result parsing, since the model is already fixed in the `WhereXXX` methods, could the result processing for transactions be automatically merged internally (based on matching UUIDs)? Additionally, since we still have `CheckOperationResults` for validation, returning `[]OperationError` from `GetSelectResults` should no longer be necessary.

The revised function signature would be:
`GetSelectResults(results []ovsdb.OperationResult, targetSlice interface{}) error`

Usage Example:

```go
var multiCondBridges []bridgeType
multiConditions := []model.Condition{
	{
		Field:    &bridgeModelInstance.Name, // Condition 1: Name == "br-sel1"
		Function: ovsdb.ConditionEqual,
		Value:    bridgeName1,
	},
	{
		Field:    &bridgeModelInstance.ExternalIDs, // Condition 2: external_ids includes {"type": "main"}
		Function: ovsdb.ConditionIncludes,
		Value:    map[string]string{"type": "main"},
	},
}

selectOpMulti, err := selectClient.WhereAny(&bridgeType{}, multiConditions...).Select()
replyMulti, err := selectClient.Transact(ctx, selectOpMulti...)
_, err = ovsdb.CheckOperationResults(replyMulti, selectOpMulti)
err = selectClient.ParseSelectResult(replyMulti, &multiCondBridges)
```


Key Changes:

- Automatic merging of transaction results by UUID internally（Even though multiple Select operations are performed, their selected columns remain consistent）.
- Removed []OperationError return from GetSelectResults since validation is handled by CheckOperationResults.
- Updated method names to reflect potential Golang conventions (e.g., ParseSelectResult instead of the original name).

Let me know if you need further refinements!

jcaamano commented:
What happens if there is multiple select operations in the same transaction on the same table? One thing is to aggregate the results of a single select operation that spans multiple operations, but I don't see that as the same thing than as aggregating the results of multiple select operations. Or am I misunderstanding your intentions?

halfcrazy commented:
> What happens if there is multiple select operations in the same transaction on the same table? One thing is to aggregate the results of a single select operation that spans multiple operations, but I don't see that as the same thing than as aggregating the results of multiple select operations. Or am I misunderstanding your intentions?

I see what you mean, indeed, there is a problem with the aggregation logic here, aggregation should not be done by default for results of multiple queries other than those introduced by WhereAny.

jcaamano commented:
So my first suggestion is for the user to pass all operations submitted to transact along with all results obtained from transact and get back a two dimensional array with the (whereany aggregated) results of each select operation. Let's focus on this option first.

halfcrazy commented:
update
```go
api
	// Select generates the OVSDB select operation based on the condition.
	// It determines the target table and columns from the condition context.
	// Returns an error if the condition was built using WhereCache.
	// It also returns a unique correlation ID that can be used to associate
	// these operations with their results.
	Select(columns ...string) ([]ovsdb.Operation, string, error)

client	
	// GetSelectResults parses the results of a transaction containing select operations
	// and populates the target slices. The targets map is keyed by the correlation ID
	// returned from the Select operation. The value is a pointer to a slice of models.
	GetSelectResults(ops []ovsdb.Operation, results []ovsdb.OperationResult, targets map[string]interface{}) error
```

jcaamano commented on README.md:
```diff
@@ -237,6 +237,112 @@ They can also be created based on a list of Conditions:
+
+// 1. Generate Ops: Use WhereAny for one or more OR conditions
+// This generates multiple operations, one for each condition
+selectOps, queryID, err := ovs.WhereAny(lsModel, cond1, cond2).Select()
+// ...
+// 2. Transact and parse
+reply, err := ovs.Transact(ctx, selectOps...)
+// GetSelectResults will parse all results associated with the queryID
+// and populate them into the single target slice.
+err = ovs.GetSelectResults(selectOps, reply, map[string]interface{}{
+	queryID: &specificSwitches,
+})
```
Why do we need to export this `queryID/CorrelationID' concept. Why isn't it just an unexported field in the OPs and only used internally?

In that way also think `GetSelectResults` should take a slice of slices (instead of a map), one slice per select operation:
```
results := [][]interface{}{}
// internally correlates using an unexported id on each op in ops
ovs.GetSelectResults(ops, results)
// this is result from first select
firstResult := results[0]
// this is result from second select
secondResult := results[1]
```

I am not sure why a user would be interested only on partial results instead of all results, so I wouldn't take it further. However, if there is a reason, then it would just be
```
// this is result from second select
ovs.GetSelectResults(ops, 1, results)
```

jcaamano replied:
> To handle the `WhereAny` case, we will merge the results for the same query ID.

But you can still do that with an unexported field in `ovsdb.Operation`?

> The two-dimensional array may be confused with the index.

I don't know what you mean, can you give me an example?

jcaamano replied:
> 2\. especially if there are pending results for another kind of table

I see this might be a concern. Let me enrich my example.

```
// select on type A
ops := client.WhereAny(&A{}, conds...).Select()
// mutate on type B
ops := append(client.WhereAny(&B{}, conds...).Mutate())
// select on type C
ops := append(client.WhereAny(&C{}, conds...).Select())
// another select on type A
ops := append(client.WhereAny(&A{}, conds...).Select())

result := client.Transact(ops...)

// get results on type A
resultsA :=[][]&A{}
ovs.GetSelectResults(ops, result, resultsA)
// results of first select on A
firstResultsA := resultsA[0]
// results of second select A
secondResultsA := resultsA[0]

// get results on type C
resultsC :=[][]&C{}
ovs.GetSelectResults(ops, result, resultsC)
// results of select on C
firstResultsC := resultsC[0]
```

Can we do this?

I guess another concept I am introducing in this example is avoiding the use of `interface{}` on `GetSelectResults` and use `m.Model` with the assumption that any given invocation is about a single type of model, just as we do in the rest of the API.

jcaamano replied:
@dave-tucker maybe you have an opinion here as well!

dave-tucker reviewed (approved):
LGTM! Please squash your commits.
I'd like @jcaamano to review too before we hit merge.

halfcrazy commented:
```
=== RUN   TestOVSIntegrationTestSuite/TestMonitorConditionIntegration
==================
WARNING: DATA RACE
Write at 0x00c001595d00 by goroutine 254:
  github.com/ovn-kubernetes/libovsdb/test/ovs.(*OVSIntegrationSuite).TestMonitorConditionIntegration()
      /home/runner/work/libovsdb/libovsdb/test/ovs/ovs_integration_test.go:825 +0xa5a
  runtime.call16()
      /home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.0.linux-amd64/src/runtime/asm_amd64.s:775 +0x42
  reflect.Value.Call()
      /home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.0.linux-amd64/src/reflect/value.go:365 +0xb5
  github.com/stretchr/testify/suite.Run.func1()
      /home/runner/go/pkg/mod/github.com/stretchr/testify@v1.10.0/suite/suite.go:202 +0x6f1
  testing.tRunner()
      /home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.0.linux-amd64/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.0.linux-amd64/src/testing/testing.go:1743 +0x44

Previous write at 0x00c001595d00 by goroutine 281:
  github.com/ovn-kubernetes/libovsdb/test/ovs.(*OVSIntegrationSuite).TestMonitorConditionIntegration.func2()
      /home/runner/work/libovsdb/libovsdb/test/ovs/ovs_integration_test.go:821 +0x144
  github.com/stretchr/testify/assert.Never.func1()
      /home/runner/go/pkg/mod/github.com/stretchr/testify@v1.10.0/assert/assertions.go:2080 +0x33

Goroutine 254 (running) created at:
  testing.(*T).Run()
      /home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.0.linux-amd64/src/testing/testing.go:1743 +0x825
  github.com/stretchr/testify/suite.runTests()
      /home/runner/go/pkg/mod/github.com/stretchr/testify@v1.10.0/suite/suite.go:247 +0x17e
  github.com/stretchr/testify/suite.Run()
      /home/runner/go/pkg/mod/github.com/stretchr/testify@v1.10.0/suite/suite.go:220 +0xa2e
  github.com/ovn-kubernetes/libovsdb/test/ovs.TestOVSIntegrationTestSuite()
      /home/runner/work/libovsdb/libovsdb/test/ovs/ovs_integration_test.go:214 +0xcb
  testing.tRunner()
      /home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.0.linux-amd64/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.0.linux-amd64/src/testing/testing.go:1743 +0x44

Goroutine 281 (finished) created at:
  github.com/stretchr/testify/assert.Never()
      /home/runner/go/pkg/mod/github.com/stretchr/testify@v1.10.0/assert/assertions.go:2080 +0x3d5
  github.com/stretchr/testify/assert.(*Assertions).Never()
      /home/runner/go/pkg/mod/github.com/stretchr/testify@v1.10.0/assert/assertion_forward.go:1009 +0xc7
  github.com/ovn-kubernetes/libovsdb/test/ovs.(*OVSIntegrationSuite).TestMonitorConditionIntegration()
      /home/runner/work/libovsdb/libovsdb/test/ovs/ovs_integration_test.go:819 +0x9d6
  runtime.call16()
      /home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.0.linux-amd64/src/runtime/asm_amd64.s:775 +0x42
  reflect.Value.Call()
      /home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.0.linux-amd64/src/reflect/value.go:365 +0xb5
  github.com/stretchr/testify/suite.Run.func1()
      /home/runner/go/pkg/mod/github.com/stretchr/testify@v1.10.0/suite/suite.go:202 +0x6f1
  testing.tRunner()
      /home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.0.linux-amd64/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.0.linux-amd64/src/testing/testing.go:1743 +0x44
==================
    testing.go:1399: race detected during execution of test
```

jcaamano commented on ovsdb/notation.go:
```diff
@@ -43,6 +43,10 @@ type Operation struct {
 	UUID      string      `json:"uuid,omitempty"`
 	UUIDName  string      `json:"uuid-name,omitempty"`
+
+	// CorrelationID is a client-side mechanism to correlate a set of operations
+	// with their results. It is not serialized.
+	CorrelationID string `json:"-"`
```
Preferably not on the type. I think a free function `func Get/SetCorrelationID(op ovsdb.Operation) string` coupled with godoc text warning of its internal nature and advising against use is good enough.

We could take it further and have an internal type that embeds Operation and adds correlationID, and type cast to it as necessary.

jcaamano commented on mapper/mapper.go:
```diff
@@ -80,6 +80,20 @@ func (m Mapper) getData(ovsData ovsdb.Row, result *Info) error {
 		}
 	}
+
+	// Explicitly handle the _uuid column after processing schema columns
+	if uuidOvsElem, uuidOk := ovsData["_uuid"]; uuidOk {
+		if uuidInfo, uuidInfoOk := uuidOvsElem.(ovsdb.UUID); uuidInfoOk {
+			// Check if the target model has a field tagged with "_uuid"
+			// The check `hasColumn` uses Metadata.Fields which is keyed by column name (tag)
+			if result.hasColumn("_uuid") {
+				// Set the field using the string value. SetField should handle it.
+				if err := result.SetField("_uuid", uuidInfo.GoUUID); err != nil {
+					return fmt.Errorf("failed to set _uuid field: %w", err)
+				}
+			}
+		}
+	}
```
For example, this has implications on the test server side implementation of Wait where before we wouldn't allow the use of the `_uuid` column but with this change we would. Although not too concerned given that it is a test server, it could be out of spec fro a real server?

What I would probably do is rename this `getData` method to `getRowData` and add a `getRowDataWithUUID` method and then let people chose what to use.

halfcrazy commented on client/api.go:
```diff
@@ -637,3 +667,84 @@ func newConditionalAPI(cache *cache.TableCache, cond Conditional, logger *logr.L
+		return nil, fmt.Errorf("Select called on API with no condition set (use Where, WhereAll, or WhereAny first)")
+	}
+
+	if _, ok := a.cond.(*predicateConditional); ok {
+		// Prevent select based on cache predicate function which cannot be translated
+		return nil, fmt.Errorf("cannot generate OVSDB select operation from a cache predicate function (WhereCache)")
+	}
```
Since the restrictions have been lifted, we can directly use `cond.generate`. There is no need to add additional constraints or create a dedicated `interface` for `WhereCache` to hide the `Select` method. The use case is still the same as you mentioned earlier, where the cache only listens to some fields and uses `Select` to complete the rest.

jcaamano commented on client/api.go:
```diff
@@ -216,13 +227,32 @@ func (a api) conditionFromFunc(predicate any) Conditional {
-		return newErrorConditional(fmt.Errorf("at least one model required"))
+	tableName := ""
+	var err error
+	// If models is empty, this is a select all on a table to be determined
+	// by the operation that uses it, e.g: List()
+	if len(models) > 0 {
+		tableName, err = a.getTableFromModel(models[0])
+		if err != nil {
+			return newErrorConditional(err)
+		}
```
hmm

OK so I might change my opinion on my original comment after thinking about this. But let's think through it all.

I would go for this option at first:
```
client.Select(model, columns..)
client.Where(model).Select(columns...)
```

Now, my original comment said that a user should be able to select all without thinking about conditions but that's not really possible: What if `model` in `client.Select(model, columns..)` carries indexes? Ignoring them would be confusing, and generating conditions from them make the user having to think about conditional search even when not using the conditional API and would also make the functionality of `client.Select(model, columns..)` and `client.Where(model).Select(columns...)` overlap as there are uses for which you could use one or the other.

Using `client.SelectAll(model, columns..)` as you propose could be an alternative, but still gives me the feeling that the model passed there is less useful that it could be.

So now I am inclined to go back to your original proposal of just having `client.Where(model).Select(columns...)` as a single clear API (with the caveat of passing an empty model to select all) but I guess I have a clearer view now why we would be doing that.

Let me know if you agree and if so, I am really sorry to the misguidance provided here.

jcaamano reviewed (commented):
Getting there.

Also missing some kind of resolution for
https://github.com/ovn-kubernetes/libovsdb/pull/399#discussion_r2282878712
https://github.com/ovn-kubernetes/libovsdb/pull/399#discussion_r2282986730

jcaamano commented on client/api.go:
```diff
@@ -637,3 +662,77 @@ func newConditionalAPI(cache *cache.TableCache, cond Conditional, logger *logr.L
+
+	// If no conditions were generated (e.g. select all), create a single
+	// operation with an empty where clause which selects all rows.
+	if len(ovsdbConditionsList) == 0 {
+		ovsdbConditionsList = append(ovsdbConditionsList, []ovsdb.Condition{})
+	}
```
nit:
```
	if len(ovsdbConditionsList) == 0 {
		ovsdbConditionsList = [][]ovsdb.Condition{[]ovsdb.Condition{}}
	}
```

jcaamano commented on client/condition.go:
```diff
@@ -246,3 +246,40 @@ func newErrorConditional(err error) Conditional {
+	return nil, fmt.Errorf("empty model conditions are not supported for data modification operations (Update, Delete, Mutate). Use specific conditions to target the rows you want to modify")
+}
+
+func (c *zeroValueConditional) Matches() (map[string]model.Model, error) {
```
I didn't realize this before but this is yet another quirk:

`Where(nonZeroModel).Select()` will lookup the cache and do a select for the results found on the cache
`Where(zeroModel).Select()` will not lookup on the cache

The other quirk is that `Where(zeroModel).List()` will error, and `Where(zeroModel).Select()` will do something.

That puts me again thinking on the `SelectAll` option you suggested before. If we did that, then our behavior across all `WhereXXX().XXXX()` would be consistent and we would require less changes on the conditional API implementation. Whereas `SelectAll(model, columns)` would require a specific implementation but at least the only user facing quirk is that we would document that model is not used to index. I am now inclined on that last option, sorry for the back and forth. Thoughts?

halfcrazy replied:
> Where(nonZeroModel).Select() will lookup the cache and do a select for the results found on the cache

We can translate the index-based condition to an equal condition. And don't use cache for consistency. 

> The other quirk is that Where(zeroModel).List() will error, and Where(zeroModel).Select() will do something.

Acknowledged. This will be different.

I thought `SelectAll(model, columns)` is clearer than the internal conditional solution.

halfcrazy commented on client/api.go:
```diff
@@ -81,6 +83,11 @@ type ConditionalAPI interface {
+	// Select returns the operations to search on the database.
+	// Depending on the Condition, it might return one or many operations.
+	// Use GetSelectResults on the results of the transaction to gather the found Models
+	Select(columns ...string) ([]ovsdb.Operation, error)
```
Note, when I implement this, looks still needs some modification to the condition to store model info since select dont require a model param.
https://github.com/ovn-kubernetes/libovsdb/pull/399/files#diff-9c46035c253e42f64c0821dd8d5dc155c0efc59a3ddc51178792da3e62e5bc63R25-R30

```
// ModelProvider is an optional interface that Conditional implementations
// can implement to provide the model instance used for field pointer resolution.
type ModelProvider interface {
	GetModel() model.Model
}
func (c *equalityConditional) GetModel() model.Model {
	if len(c.models) > 0 {
		return c.models[0]
	}
	return nil
}
func (c *explicitConditional) GetModel() model.Model {
	return c.model
}
func (c *predicateConditional) GetModel() model.Model {
	return c.model
}
```

halfcrazy commented on client/api.go:
```diff
@@ -637,3 +652,139 @@ func newConditionalAPI(cache *cache.TableCache, cond Conditional, logger *logr.L
+}
+
+// SelectAll selects all rows from a table, with optional column filtering.
+func (a api) SelectAll(m model.Model, fields ...any) ([]ovsdb.Operation, error) {
```
emm, so just one function. Combine `select` and `selectall` always receive a model parameter, when `conditionalAPI` `Select` checks the first m parameter is empty and consistent with the conditional model.

halfcrazy commented on client/api.go:
```diff
@@ -637,3 +652,139 @@ func newConditionalAPI(cache *cache.TableCache, cond Conditional, logger *logr.L
+
+// getColumnsToSelect is a helper function that determines which columns to select
+// based on a model and a list of field pointers.
+func (a api) getColumnsToSelect(m model.Model, tableSchema *ovsdb.TableSchema, tableName string, fields ...any) ([]string, error) {
```
Will mutate update wait using `func (a api) getColumns(m model.Model, fields ...any) ([]string, error) `  ? It doesn't expose mutable/immutable information.

halfcrazy commented:
Thrilled that we’re finally at the merge stage for this PR! I know this one took a few rounds of review, but I genuinely valued every comment you shared – it not only made the code more robust but also taught me a lot about api design.

Thanks for taking the time to walk through this with me step by step. Couldn’t have gotten here without your guidance!
